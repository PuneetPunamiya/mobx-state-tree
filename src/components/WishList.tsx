import React from 'react'
import {observer} from 'mobx-react'

import WishListItemView from './WishListItemView'
// import {WishList} from '../models/WishList'

const WishListView = (props: any) => {
    const {wishlist} = props
    return (
        <div>
            <ul>
                {wishlist.items.map((item: any, idx: any) => <WishListItemView key={idx} {...item} />)}
            </ul>
            Total: {wishlist.totalPrice}
        </div>
    )
}

export default observer(WishListView)