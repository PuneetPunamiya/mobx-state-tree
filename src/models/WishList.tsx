import {types} from 'mobx-state-tree'

export const WishListItem = types.model({
    name: types.string,
    price: types.number
})
    .actions(self => ({
        changeName(newName: string) {
            self.name = newName
        },
        changePrice(newPrice: number) {
            self.price = newPrice
        }
    }))

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), [])
})
    .actions(self => ({
        add(item: any) {
            self.items.push(item)
        }
    }))
    .views(self => ({
        get totalPrice() {
            return self.items.reduce((sum, entry) =>
                sum + entry.price, 0
            )
        }
    }))