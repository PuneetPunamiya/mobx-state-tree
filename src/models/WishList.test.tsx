import {WishListItem, WishList} from './WishList'
import {getSnapshot, onSnapshot, onPatch} from 'mobx-state-tree'
import {reaction} from 'mobx'

it("can create an instance of a model", () => {
    const items = WishListItem.create({
        "name": "abc",
        "price": 123,
    })

    expect(items.price).toBe(123)
    items.changeName("Puneet")
    expect(items.name).toBe("Puneet")
})

it("can create a wishlist", () => {
    const list = WishList.create({
        items: [
            {
                "name": "abc",
                "price": 123,
            }
        ]
    })

    expect(list.items.length).toBe(1)
})

it("can add new items", () => {
    const list = WishList.create()

    list.add(({
        "name": "qwe",
        "price": 345,
    }))

    const states: any = []
    onSnapshot(list, snapshot => {
        states.push(snapshot)
    })

    const patches: any = []
    onPatch(list, patch => {
        patches.push(patch)
    })

    expect(list.items.length).toBe(1)
    list.items[0].changeName("asd")
    expect(list.items[0].name).toBe("asd")

    expect(getSnapshot(list)).toMatchSnapshot()

    expect(states).toMatchSnapshot()

    expect(patches).toMatchSnapshot()
})

it("can calculate the total price of the wishlist", () => {
    const list = WishList.create({
        items: [
            {
                "name": "abc",
                "price": 123,
            },
            {
                "name": "qwe",
                "price": 123,
            }
        ]
    })

    let changed = 0
    reaction(() => list.totalPrice, () => changed++)
    console.log(list.totalPrice)

    //Only if the price is changed then the value of 'changed' will increase to 1
    list.items[0].changePrice(10)

    expect(changed).toBe(1)
    expect(list.totalPrice).toBe(133)

    list.items.map((item: any) => {console.log(item)})

})