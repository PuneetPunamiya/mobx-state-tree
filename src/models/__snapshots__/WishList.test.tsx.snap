// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "name": "asd",
      "price": 345,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "name": "asd",
        "price": 345,
      },
    ],
  },
]
`;

exports[`can add new items 3`] = `
Array [
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "asd",
  },
]
`;
